grug/malachite/src/actors/host/host.rs

```diff
@@ -0,0 +1,560 @@
+            let block_hash = state.finalize_block(&pre_block)?;
+
+            pre_block.into_block(block_hash)
+        };
+
+        let proposal_data = block.as_proposal_data();
+
+        let stream_id = state.stream_id();
+
+        // Store undecided proposal
+        let proposed_value = ProposedValue {
+            height,
+            round,
+            valid_round: Round::Nil,
+            proposer: self.address,
+            value: <ctx!(Value)>::new(proposal_data.block.block_hash()),
+            validity: Validity::Valid,
+        };
+
+        UNDECIDED_PROPOSALS.save(
+            state,
+            (*height, round.as_i64(), proposal_data.block.block_hash()),
+            &proposed_value,
+        )?;
+
+        // Store undecided block
+        UNDECIDED_BLOCK.save(state, proposal_data.block.block_hash(), &block)?;
+
+        let value = LocallyProposedValue::new(
+            height,
+            round,
+            <ctx!(Value)>::new(proposal_data.block.block_hash()),
+        );
+
+        // Stream parts to consensus
+        self.stream_data(stream_id, proposal_data, true).await?;
+
+        // Return the proposed value
+        reply_to.send(value)?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument("received_proposal_part", skip_all)]
+    async fn received_proposal_part(
+        &self,
+        state: &mut State,
+        part: StreamMessage<ctx!(ProposalPart)>,
+        reply_to: RpcReplyPort<ProposedValue<Context>>,
+    ) -> ActorResult<()> {
+        // Since we are using a single part proposal, we only need data
+        let Some(ProposalData { block, valid_round }) = part.content.into_data() else {
+            return Ok(());
+        };
+
+        info!(height = %block.height, round = %block.round, "All parts received");
+
+        // Run FinalizeBlock
+        let block_hash = state.finalize_block(&block)?;
+
+        let value = ProposedValue {
+            height: block.height,
+            round: block.round,
+            valid_round,
+            proposer: block.proposer,
+            // TODO: Is this correct? If block hashes are different, which one we should use?
+            value: <ctx!(Value)>::new(block.block_hash()),
+            validity: Validity::from_bool(block.block_hash() == block_hash),
+        };
+
+        if value.validity == Validity::Valid {
+            info!(block_hash = %block_hash, "Block hash matches");
+            // Store undecided proposal
+            UNDECIDED_PROPOSALS.save(
+                state,
+                (*block.height, block.round.as_i64(), block.block_hash()),
+                &value,
+            )?;
+
+            // Store undecided block
+            UNDECIDED_BLOCK.save(state, block.block_hash(), &block)?;
+        } else {
+            warn!(block_hash = %block_hash, proposal_block_hash = %block.block_hash(), "Block hash mismatch");
+        }
+
+        // TODO: If block_hash is different from the proposed value, what we should do?
```

Comment:
One reason to store is to handle restarts. The other is to better handle non-determinism and in general to have same code for both valid and invalid proposals up to commit. See comments on `decided()`